### PR TITLE
Rework the keep item completely

### PR DESCRIPTION
**Pull request by Raycoms**
# Changes proposed in this pull request:
- Completely reworks dumping and item keeping
- Workers now keep predicates
-

Review please

### CODE REVIEW

File: src/main/java/com/minecolonies/coremod/colony/buildings/BuildingLumberjack.java [hunk 1]
```diff
@@ -23,9 +23,9 @@
 import net.minecraftforge.fml.common.network.ByteBufUtils;
 import net.minecraftforge.oredict.OreDictionary;
 import org.jetbrains.annotations.NotNull;
-import org.jetbrains.annotations.Nullable;
 
 import java.util.*;
+import java.util.function.Predicate;
```
Comment (MinecoloniesSonar) [inline src/main/java/com/minecolonies/coremod/colony/buildings/BuildingLumberjack.java]: MINOR Remove this unused import 'java.util.function.Predicate'. [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AUselessImportCheck)

File: src/main/java/com/minecolonies/coremod/colony/buildings/AbstractBuildingGuards.java [hunk 2]
```diff
@@ -14,6 +16,8 @@
 import io.netty.buffer.ByteBuf;
 import net.minecraft.entity.SharedMonsterAttributes;
 import net.minecraft.entity.player.EntityPlayer;
+import net.minecraft.init.Blocks;
```
Comment (MinecoloniesSonar) [inline src/main/java/com/minecolonies/coremod/colony/buildings/AbstractBuildingGuards.java]: MINOR Remove this unused import 'net.minecraft.init.Blocks'. [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AUselessImportCheck)

File: src/main/java/com/minecolonies/coremod/colony/buildings/AbstractBuildingGuards.java [hunk 3]
```diff
@@ -1,7 +1,9 @@
 package com.minecolonies.coremod.colony.buildings;
 
 import com.minecolonies.api.util.BlockPosUtil;
+import com.minecolonies.api.util.InventoryUtils;
```
Comment (MinecoloniesSonar) [inline src/main/java/com/minecolonies/coremod/colony/buildings/AbstractBuildingGuards.java]: MINOR Remove this unused import 'com.minecolonies.api.util.InventoryUtils'. [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AUselessImportCheck)

File: src/main/java/com/minecolonies/coremod/colony/buildings/BuildingBaker.java [hunk 4]
```diff
@@ -29,6 +28,7 @@
 import org.jetbrains.annotations.NotNull;
 
 import java.util.*;
+import java.util.function.Predicate;
```
Comment (MinecoloniesSonar) [inline src/main/java/com/minecolonies/coremod/colony/buildings/BuildingBaker.java]: MINOR Remove this unused import 'java.util.function.Predicate'. [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AUselessImportCheck)

File: src/main/java/com/minecolonies/coremod/entity/ai/citizen/deliveryman/EntityAIWorkDeliveryman.java [hunk 5]
```diff
@@ -29,6 +29,7 @@
 import java.util.Collection;
 import java.util.List;
 import java.util.Map;
+import java.util.function.Predicate;
```
Comment (MinecoloniesSonar) [inline src/main/java/com/minecolonies/coremod/entity/ai/citizen/deliveryman/EntityAIWorkDeliveryman.java]: MINOR Remove this unused import 'java.util.function.Predicate'. [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AUselessImportCheck)

File: src/main/java/com/minecolonies/coremod/entity/ai/citizen/deliveryman/EntityAIWorkDeliveryman.java [hunk 6]
```diff
@@ -209,7 +210,8 @@ private boolean gatherFromBuilding(@NotNull final AbstractBuilding building)
         }
 
         if (workerRequiresItem(building, stack, alreadyKept)
-                || (building instanceof BuildingHome && stack.getItem() instanceof ItemFood))
+                || (building instanceof BuildingHome
+                && stack.getItem() instanceof ItemFood))
```
Comment (MinecoloniesSonar) [inline src/main/java/com/minecolonies/coremod/entity/ai/citizen/deliveryman/EntityAIWorkDeliveryman.java:213]: MAJOR A "NullPointerException" could be thrown; "stack" is nullable here. [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS2259)

File: src/main/java/com/minecolonies/coremod/colony/buildings/BuildingMiner.java [hunk 7]
```diff
@@ -30,6 +30,7 @@
 import java.util.HashMap;
 import java.util.List;
 import java.util.Map;
```
Comment (MinecoloniesSonar) [inline src/main/java/com/minecolonies/coremod/colony/buildings/BuildingMiner.java]: MINOR Remove this unused import 'java.util.Map'. [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AUselessImportCheck)

File: src/main/java/com/minecolonies/coremod/colony/buildings/BuildingMiner.java [hunk 8]
```diff
@@ -30,6 +30,7 @@
...
+import java.util.function.Predicate;
```
Comment (MinecoloniesSonar) [inline src/main/java/com/minecolonies/coremod/colony/buildings/BuildingMiner.java]: MINOR Remove this unused import 'java.util.function.Predicate'. [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AUselessImportCheck)

File: src/main/java/com/minecolonies/coremod/colony/buildings/BuildingMiner.java [hunk 9]
```diff
@@ -30,6 +30,7 @@
 import java.util.HashMap;
```
Comment (MinecoloniesSonar) [inline src/main/java/com/minecolonies/coremod/colony/buildings/BuildingMiner.java]: MINOR Remove this unused import 'java.util.HashMap'. [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AUselessImportCheck)

File: src/main/java/com/minecolonies/coremod/entity/ai/basic/AbstractEntityAIBasic.java [hunk 10]
```diff
@@ -856,121 +847,18 @@ private void clearActionsDone()
...
-            return ItemStackUtils.EMPTY;
-        }
-        alreadyKept.put(tempStorage, shouldKeep.get(tempStorage));
-        final int dump = tempStorage.getAmount() + amountKept - shouldKeep.get(tempStorage);
-
-        //Create tempStack with the amount of items that should be dumped.
-        return new ItemStack(tempStorage.getItem(), dump, tempStorage.getDamageValue());
+                || InventoryFunctions.matchFirstInHandlerWithAction(new InvWrapper(worker.getInventoryCitizen()),
+                itemStack -> !ItemStackUtils.isEmpty(itemStack) && !AbstractBuilding.buildingRequiresCertainAmountOfItem(buildingWorker, itemStack, alreadyKept),
+                (handler, slot) -> InventoryUtils.transferItemStackIntoNextFreeSlotInItemHandlers(new InvWrapper(worker.getInventoryCitizen()), slot, new InvWrapper(buildingWorker.getTileEntity()))
```
Comment (MinecoloniesSonar) [inline src/main/java/com/minecolonies/coremod/entity/ai/basic/AbstractEntityAIBasic.java]: MAJOR Split this 197 characters long line (which is greater than 180 authorized). [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS00103)

**Comment by MinecoloniesSonar** [comment]
SonarQube analysis reported 89 issues
* ![MAJOR][MAJOR] 19 major
* ![MINOR][MINOR] 69 minor
* ![INFO][INFO] 1 info

Watch the comments in this conversation to review them.

#### Top 10 extra issues

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. ![MAJOR][MAJOR] [AbstractBuilding.java#L53](https://github.com/Minecolonies/minecolonies/blob/0d83b52da87a432ad13ce746700c369a14b5e0f2/src/main/java/com/minecolonies/coremod/colony/buildings/AbstractBuilding.java#L53): Split this class into smaller and more specialized ones to reduce its dependencies on other classes from 39 to the maximum authorized 30 or less. [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1200)
1. ![MAJOR][MAJOR] [AbstractBuilding.java#L53](https://github.com/Minecolonies/minecolonies/blob/0d83b52da87a432ad13ce746700c369a14b5e0f2/src/main/java/com/minecolonies/coremod/colony/buildings/AbstractBuilding.java#L53): class "AbstractBuilding" has 68 methods, which is greater than the 35 authorized. Split it into smaller classes. [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1448)
1. ![MAJOR][MAJOR] [AbstractBuilding.java#L1353](https://github.com/Minecolonies/minecolonies/blob/0d83b52da87a432ad13ce746700c369a14b5e0f2/src/main/java/com/minecolonies/coremod/colony/buildings/AbstractBuilding.java#L1353): Reduce this class from 75 to the maximum allowed 30 or externalize it in a public class. [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS2972)
1. ![MAJOR][MAJOR] [AbstractBuildingGuards.java#L571](https://github.com/Minecolonies/minecolonies/blob/0d83b52da87a432ad13ce746700c369a14b5e0f2/src/main/java/com/minecolonies/coremod/colony/buildings/AbstractBuildingGuards.java#L571): Reduce this class from 58 to the maximum allowed 30 or externalize it in a public class. [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS2972)
1. ![MAJOR][MAJOR] [AbstractBuildingWorker.java#L179](https://github.com/Minecolonies/minecolonies/blob/0d83b52da87a432ad13ce746700c369a14b5e0f2/src/main/java/com/minecolonies/coremod/colony/buildings/AbstractBuildingWorker.java#L179): Either log or rethrow this exception. [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1166)
1. ![MAJOR][MAJOR] [AbstractBuildingWorker.java#L330](https://github.com/Minecolonies/minecolonies/blob/0d83b52da87a432ad13ce746700c369a14b5e0f2/src/main/java/com/minecolonies/coremod/colony/buildings/AbstractBuildingWorker.java#L330): Reduce this class from 52 to the maximum allowed 30 or externalize it in a public class. [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS2972)
1. ![MAJOR][MAJOR] [BuildingFarmer.java#L536](https://github.com/Minecolonies/minecolonies/blob/0d83b52da87a432ad13ce746700c369a14b5e0f2/src/main/java/com/minecolonies/coremod/colony/buildings/BuildingFarmer.java#L536): Reduce this class from 86 to the maximum allowed 30 or externalize it in a public class. [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS2972)
1. ![MAJOR][MAJOR] [BuildingLumberjack.java#L216](https://github.com/Minecolonies/minecolonies/blob/0d83b52da87a432ad13ce746700c369a14b5e0f2/src/main/java/com/minecolonies/coremod/colony/buildings/BuildingLumberjack.java#L216): Reduce this class from 73 to the maximum allowed 30 or externalize it in a public class. [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS2972)
1. ![MAJOR][MAJOR] [BuildingMiner.java#L462](https://github.com/Minecolonies/minecolonies/blob/0d83b52da87a432ad13ce746700c369a14b5e0f2/src/main/java/com/minecolonies/coremod/colony/buildings/BuildingMiner.java#L462): Assign this magic number 3 to a well-named constant, and use the constant instead. [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS109)
1. ![MAJOR][MAJOR] [BuildingMiner.java#L639](https://github.com/Minecolonies/minecolonies/blob/0d83b52da87a432ad13ce746700c369a14b5e0f2/src/main/java/com/minecolonies/coremod/colony/buildings/BuildingMiner.java#L639): Reduce this class from 39 to the maximum allowed 30 or externalize it in a public class. [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS2972)

[INFO]: https://sonarsource.github.io/sonar-github/severity-info.png 'Severity: INFO'
[MAJOR]: https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR'
[MINOR]: https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR'

**Comment by MinecoloniesSonar** [comment]
SonarQube analysis reported 83 issues
* ![MAJOR][MAJOR] 22 major
* ![MINOR][MINOR] 59 minor
* ![INFO][INFO] 2 info

Watch the comments in this conversation to review them.

#### Top 10 extra issues

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. ![MAJOR][MAJOR] [AbstractBuilding.java#L53](https://github.com/Minecolonies/minecolonies/blob/afc6b3b64743d6d01638915461bdd258c780300f/src/main/java/com/minecolonies/coremod/colony/buildings/AbstractBuilding.java#L53): Split this class into smaller and more specialized ones to reduce its dependencies on other classes from 39 to the maximum authorized 30 or less. [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1200)
1. ![MAJOR][MAJOR] [AbstractBuilding.java#L53](https://github.com/Minecolonies/minecolonies/blob/afc6b3b64743d6d01638915461bdd258c780300f/src/main/java/com/minecolonies/coremod/colony/buildings/AbstractBuilding.java#L53): class "AbstractBuilding" has 68 methods, which is greater than the 35 authorized. Split it into smaller classes. [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1448)
1. ![MAJOR][MAJOR] [AbstractBuilding.java#L1353](https://github.com/Minecolonies/minecolonies/blob/afc6b3b64743d6d01638915461bdd258c780300f/src/main/java/com/minecolonies/coremod/colony/buildings/AbstractBuilding.java#L1353): Reduce this class from 75 to the maximum allowed 30 or externalize it in a public class. [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS2972)
1. ![MAJOR][MAJOR] [AbstractBuildingGuards.java#L569](https://github.com/Minecolonies/minecolonies/blob/afc6b3b64743d6d01638915461bdd258c780300f/src/main/java/com/minecolonies/coremod/colony/buildings/AbstractBuildingGuards.java#L569): Reduce this class from 58 to the maximum allowed 30 or externalize it in a public class. [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS2972)
1. ![MAJOR][MAJOR] [AbstractBuildingWorker.java#L178](https://github.com/Minecolonies/minecolonies/blob/afc6b3b64743d6d01638915461bdd258c780300f/src/main/java/com/minecolonies/coremod/colony/buildings/AbstractBuildingWorker.java#L178): Either log or rethrow this exception. [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1166)
1. ![MAJOR][MAJOR] [AbstractBuildingWorker.java#L329](https://github.com/Minecolonies/minecolonies/blob/afc6b3b64743d6d01638915461bdd258c780300f/src/main/java/com/minecolonies/coremod/colony/buildings/AbstractBuildingWorker.java#L329): Reduce this class from 52 to the maximum allowed 30 or externalize it in a public class. [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS2972)
1. ![MAJOR][MAJOR] [BuildingFarmer.java#L535](https://github.com/Minecolonies/minecolonies/blob/afc6b3b64743d6d01638915461bdd258c780300f/src/main/java/com/minecolonies/coremod/colony/buildings/BuildingFarmer.java#L535): Reduce this class from 86 to the maximum allowed 30 or externalize it in a public class. [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS2972)
1. ![MAJOR][MAJOR] [BuildingLumberjack.java#L215](https://github.com/Minecolonies/minecolonies/blob/afc6b3b64743d6d01638915461bdd258c780300f/src/main/java/com/minecolonies/coremod/colony/buildings/BuildingLumberjack.java#L215): Reduce this class from 73 to the maximum allowed 30 or externalize it in a public class. [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS2972)
1. ![MAJOR][MAJOR] [BuildingMiner.java#L458](https://github.com/Minecolonies/minecolonies/blob/afc6b3b64743d6d01638915461bdd258c780300f/src/main/java/com/minecolonies/coremod/colony/buildings/BuildingMiner.java#L458): Assign this magic number 3 to a well-named constant, and use the constant instead. [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS109)
1. ![MAJOR][MAJOR] [BuildingMiner.java#L635](https://github.com/Minecolonies/minecolonies/blob/afc6b3b64743d6d01638915461bdd258c780300f/src/main/java/com/minecolonies/coremod/colony/buildings/BuildingMiner.java#L635): Reduce this class from 39 to the maximum allowed 30 or externalize it in a public class. [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS2972)

[INFO]: https://sonarsource.github.io/sonar-github/severity-info.png 'Severity: INFO'
[MAJOR]: https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR'
[MINOR]: https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR'

File: src/main/java/com/minecolonies/coremod/entity/ai/citizen/deliveryman/EntityAIWorkDeliveryman.java (hunk 6 above)
Comment (Raycoms) [reply src/main/java/com/minecolonies/coremod/entity/ai/citizen/deliveryman/EntityAIWorkDeliveryman.java:213]: You're a lier Sonar, you should be ashamed of yourself!

**Comment by MinecoloniesSonar** [comment]
SonarQube analysis reported 83 issues
* ![MAJOR][MAJOR] 22 major
* ![MINOR][MINOR] 59 minor
* ![INFO][INFO] 2 info

Watch the comments in this conversation to review them.

#### Top 10 extra issues

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. ![MAJOR][MAJOR] [AbstractBuilding.java#L53](https://github.com/Minecolonies/minecolonies/blob/5ce2d7ea048e1eebe7b9fa20202185a20759142d/src/main/java/com/minecolonies/coremod/colony/buildings/AbstractBuilding.java#L53): Split this class into smaller and more specialized ones to reduce its dependencies on other classes from 39 to the maximum authorized 30 or less. [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1200)
1. ![MAJOR][MAJOR] [AbstractBuilding.java#L53](https://github.com/Minecolonies/minecolonies/blob/5ce2d7ea048e1eebe7b9fa20202185a20759142d/src/main/java/com/minecolonies/coremod/colony/buildings/AbstractBuilding.java#L53): class "AbstractBuilding" has 68 methods, which is greater than the 35 authorized. Split it into smaller classes. [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1448)
1. ![MAJOR][MAJOR] [AbstractBuilding.java#L1353](https://github.com/Minecolonies/minecolonies/blob/5ce2d7ea048e1eebe7b9fa20202185a20759142d/src/main/java/com/minecolonies/coremod/colony/buildings/AbstractBuilding.java#L1353): Reduce this class from 75 to the maximum allowed 30 or externalize it in a public class. [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS2972)
1. ![MAJOR][MAJOR] [AbstractBuildingGuards.java#L569](https://github.com/Minecolonies/minecolonies/blob/5ce2d7ea048e1eebe7b9fa20202185a20759142d/src/main/java/com/minecolonies/coremod/colony/buildings/AbstractBuildingGuards.java#L569): Reduce this class from 58 to the maximum allowed 30 or externalize it in a public class. [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS2972)
1. ![MAJOR][MAJOR] [AbstractBuildingWorker.java#L178](https://github.com/Minecolonies/minecolonies/blob/5ce2d7ea048e1eebe7b9fa20202185a20759142d/src/main/java/com/minecolonies/coremod/colony/buildings/AbstractBuildingWorker.java#L178): Either log or rethrow this exception. [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1166)
1. ![MAJOR][MAJOR] [AbstractBuildingWorker.java#L329](https://github.com/Minecolonies/minecolonies/blob/5ce2d7ea048e1eebe7b9fa20202185a20759142d/src/main/java/com/minecolonies/coremod/colony/buildings/AbstractBuildingWorker.java#L329): Reduce this class from 52 to the maximum allowed 30 or externalize it in a public class. [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS2972)
1. ![MAJOR][MAJOR] [BuildingFarmer.java#L535](https://github.com/Minecolonies/minecolonies/blob/5ce2d7ea048e1eebe7b9fa20202185a20759142d/src/main/java/com/minecolonies/coremod/colony/buildings/BuildingFarmer.java#L535): Reduce this class from 86 to the maximum allowed 30 or externalize it in a public class. [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS2972)
1. ![MAJOR][MAJOR] [BuildingLumberjack.java#L215](https://github.com/Minecolonies/minecolonies/blob/5ce2d7ea048e1eebe7b9fa20202185a20759142d/src/main/java/com/minecolonies/coremod/colony/buildings/BuildingLumberjack.java#L215): Reduce this class from 73 to the maximum allowed 30 or externalize it in a public class. [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS2972)
1. ![MAJOR][MAJOR] [BuildingMiner.java#L458](https://github.com/Minecolonies/minecolonies/blob/5ce2d7ea048e1eebe7b9fa20202185a20759142d/src/main/java/com/minecolonies/coremod/colony/buildings/BuildingMiner.java#L458): Assign this magic number 3 to a well-named constant, and use the constant instead. [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS109)
1. ![MAJOR][MAJOR] [BuildingMiner.java#L635](https://github.com/Minecolonies/minecolonies/blob/5ce2d7ea048e1eebe7b9fa20202185a20759142d/src/main/java/com/minecolonies/coremod/colony/buildings/BuildingMiner.java#L635): Reduce this class from 39 to the maximum allowed 30 or externalize it in a public class. [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS2972)

[INFO]: https://sonarsource.github.io/sonar-github/severity-info.png 'Severity: INFO'
[MAJOR]: https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR'
[MINOR]: https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR'

File: src/main/java/com/minecolonies/coremod/entity/ai/basic/AbstractEntityAIBasic.java [hunk 11]
```diff
@@ -4,6 +4,7 @@
 import com.minecolonies.api.util.*;
 import com.minecolonies.api.util.constant.IToolType;
 import com.minecolonies.api.util.constant.ToolType;
+import com.minecolonies.coremod.colony.buildings.AbstractBuilding;
```
Comment (MinecoloniesSonar) [inline src/main/java/com/minecolonies/coremod/entity/ai/basic/AbstractEntityAIBasic.java]: MINOR Remove this unused import 'com.minecolonies.coremod.colony.buildings.AbstractBuilding'. [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AUselessImportCheck)

**Comment by MinecoloniesSonar** [comment]
SonarQube analysis reported 84 issues
* ![MAJOR][MAJOR] 22 major
* ![MINOR][MINOR] 60 minor
* ![INFO][INFO] 2 info

Watch the comments in this conversation to review them.

#### Top 10 extra issues

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. ![MAJOR][MAJOR] [AbstractBuilding.java#L53](https://github.com/Minecolonies/minecolonies/blob/865aca56deef54270a77b6ca2e4f8a217ca2f109/src/main/java/com/minecolonies/coremod/colony/buildings/AbstractBuilding.java#L53): Split this class into smaller and more specialized ones to reduce its dependencies on other classes from 39 to the maximum authorized 30 or less. [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1200)
1. ![MAJOR][MAJOR] [AbstractBuilding.java#L53](https://github.com/Minecolonies/minecolonies/blob/865aca56deef54270a77b6ca2e4f8a217ca2f109/src/main/java/com/minecolonies/coremod/colony/buildings/AbstractBuilding.java#L53): class "AbstractBuilding" has 68 methods, which is greater than the 35 authorized. Split it into smaller classes. [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1448)
1. ![MAJOR][MAJOR] [AbstractBuilding.java#L1353](https://github.com/Minecolonies/minecolonies/blob/865aca56deef54270a77b6ca2e4f8a217ca2f109/src/main/java/com/minecolonies/coremod/colony/buildings/AbstractBuilding.java#L1353): Reduce this class from 75 to the maximum allowed 30 or externalize it in a public class. [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS2972)
1. ![MAJOR][MAJOR] [AbstractBuildingGuards.java#L569](https://github.com/Minecolonies/minecolonies/blob/865aca56deef54270a77b6ca2e4f8a217ca2f109/src/main/java/com/minecolonies/coremod/colony/buildings/AbstractBuildingGuards.java#L569): Reduce this class from 58 to the maximum allowed 30 or externalize it in a public class. [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS2972)
1. ![MAJOR][MAJOR] [AbstractBuildingWorker.java#L178](https://github.com/Minecolonies/minecolonies/blob/865aca56deef54270a77b6ca2e4f8a217ca2f109/src/main/java/com/minecolonies/coremod/colony/buildings/AbstractBuildingWorker.java#L178): Either log or rethrow this exception. [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1166)
1. ![MAJOR][MAJOR] [AbstractBuildingWorker.java#L329](https://github.com/Minecolonies/minecolonies/blob/865aca56deef54270a77b6ca2e4f8a217ca2f109/src/main/java/com/minecolonies/coremod/colony/buildings/AbstractBuildingWorker.java#L329): Reduce this class from 52 to the maximum allowed 30 or externalize it in a public class. [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS2972)
1. ![MAJOR][MAJOR] [BuildingFarmer.java#L535](https://github.com/Minecolonies/minecolonies/blob/865aca56deef54270a77b6ca2e4f8a217ca2f109/src/main/java/com/minecolonies/coremod/colony/buildings/BuildingFarmer.java#L535): Reduce this class from 86 to the maximum allowed 30 or externalize it in a public class. [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS2972)
1. ![MAJOR][MAJOR] [BuildingLumberjack.java#L215](https://github.com/Minecolonies/minecolonies/blob/865aca56deef54270a77b6ca2e4f8a217ca2f109/src/main/java/com/minecolonies/coremod/colony/buildings/BuildingLumberjack.java#L215): Reduce this class from 73 to the maximum allowed 30 or externalize it in a public class. [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS2972)
1. ![MAJOR][MAJOR] [BuildingMiner.java#L458](https://github.com/Minecolonies/minecolonies/blob/865aca56deef54270a77b6ca2e4f8a217ca2f109/src/main/java/com/minecolonies/coremod/colony/buildings/BuildingMiner.java#L458): Assign this magic number 3 to a well-named constant, and use the constant instead. [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS109)
1. ![MAJOR][MAJOR] [BuildingMiner.java#L635](https://github.com/Minecolonies/minecolonies/blob/865aca56deef54270a77b6ca2e4f8a217ca2f109/src/main/java/com/minecolonies/coremod/colony/buildings/BuildingMiner.java#L635): Reduce this class from 39 to the maximum allowed 30 or externalize it in a public class. [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS2972)

[INFO]: https://sonarsource.github.io/sonar-github/severity-info.png 'Severity: INFO'
[MAJOR]: https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR'
[MINOR]: https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR'

**Comment by MinecoloniesSonar** [comment]
SonarQube analysis reported 83 issues
* ![MAJOR][MAJOR] 22 major
* ![MINOR][MINOR] 59 minor
* ![INFO][INFO] 2 info

Watch the comments in this conversation to review them.

#### Top 10 extra issues

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. ![MAJOR][MAJOR] [AbstractBuilding.java#L53](https://github.com/Minecolonies/minecolonies/blob/90b455918d415c12a1ffb0dee0c96dad1aba4972/src/main/java/com/minecolonies/coremod/colony/buildings/AbstractBuilding.java#L53): Split this class into smaller and more specialized ones to reduce its dependencies on other classes from 39 to the maximum authorized 30 or less. [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1200)
1. ![MAJOR][MAJOR] [AbstractBuilding.java#L53](https://github.com/Minecolonies/minecolonies/blob/90b455918d415c12a1ffb0dee0c96dad1aba4972/src/main/java/com/minecolonies/coremod/colony/buildings/AbstractBuilding.java#L53): class "AbstractBuilding" has 68 methods, which is greater than the 35 authorized. Split it into smaller classes. [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1448)
1. ![MAJOR][MAJOR] [AbstractBuilding.java#L1352](https://github.com/Minecolonies/minecolonies/blob/90b455918d415c12a1ffb0dee0c96dad1aba4972/src/main/java/com/minecolonies/coremod/colony/buildings/AbstractBuilding.java#L1352): Reduce this class from 75 to the maximum allowed 30 or externalize it in a public class. [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS2972)
1. ![MAJOR][MAJOR] [AbstractBuildingGuards.java#L569](https://github.com/Minecolonies/minecolonies/blob/90b455918d415c12a1ffb0dee0c96dad1aba4972/src/main/java/com/minecolonies/coremod/colony/buildings/AbstractBuildingGuards.java#L569): Reduce this class from 58 to the maximum allowed 30 or externalize it in a public class. [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS2972)
1. ![MAJOR][MAJOR] [AbstractBuildingWorker.java#L178](https://github.com/Minecolonies/minecolonies/blob/90b455918d415c12a1ffb0dee0c96dad1aba4972/src/main/java/com/minecolonies/coremod/colony/buildings/AbstractBuildingWorker.java#L178): Either log or rethrow this exception. [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1166)
1. ![MAJOR][MAJOR] [AbstractBuildingWorker.java#L329](https://github.com/Minecolonies/minecolonies/blob/90b455918d415c12a1ffb0dee0c96dad1aba4972/src/main/java/com/minecolonies/coremod/colony/buildings/AbstractBuildingWorker.java#L329): Reduce this class from 52 to the maximum allowed 30 or externalize it in a public class. [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS2972)
1. ![MAJOR][MAJOR] [BuildingFarmer.java#L535](https://github.com/Minecolonies/minecolonies/blob/90b455918d415c12a1ffb0dee0c96dad1aba4972/src/main/java/com/minecolonies/coremod/colony/buildings/BuildingFarmer.java#L535): Reduce this class from 86 to the maximum allowed 30 or externalize it in a public class. [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS2972)
1. ![MAJOR][MAJOR] [BuildingLumberjack.java#L215](https://github.com/Minecolonies/minecolonies/blob/90b455918d415c12a1ffb0dee0c96dad1aba4972/src/main/java/com/minecolonies/coremod/colony/buildings/BuildingLumberjack.java#L215): Reduce this class from 73 to the maximum allowed 30 or externalize it in a public class. [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS2972)
1. ![MAJOR][MAJOR] [BuildingMiner.java#L458](https://github.com/Minecolonies/minecolonies/blob/90b455918d415c12a1ffb0dee0c96dad1aba4972/src/main/java/com/minecolonies/coremod/colony/buildings/BuildingMiner.java#L458): Assign this magic number 3 to a well-named constant, and use the constant instead. [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS109)
1. ![MAJOR][MAJOR] [BuildingMiner.java#L635](https://github.com/Minecolonies/minecolonies/blob/90b455918d415c12a1ffb0dee0c96dad1aba4972/src/main/java/com/minecolonies/coremod/colony/buildings/BuildingMiner.java#L635): Reduce this class from 39 to the maximum allowed 30 or externalize it in a public class. [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS2972)

[INFO]: https://sonarsource.github.io/sonar-github/severity-info.png 'Severity: INFO'
[MAJOR]: https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR'
[MINOR]: https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR'

**Review by marvin-bitterlich — APPROVED** [review]

**Review by Asherslab — APPROVED** [review]
Looks good to me! :smile:

**Pull request merged**